tarodb/assets_common.go

```diff
@@ -334,3 +337,81 @@ func fetchGenesis(ctx context.Context, q FetchGenesisStore,
 		Type:         asset.Type(gen.AssetType),
 	}, nil
 }
+
+type GroupStore interface {
+	FetchGenesisStore
+
+	// AllGroups fetches all asset groups on disk.
+	AllGroups(context.Context) ([]sqlc.KeyGroupInfoView, error)
+}
+
+// FetchAllGroups fetches all asset groups on disk.
+func fetchAllGroups(ctx context.Context, q GroupStore) ([]*asset.AssetGroup, error) {
+
+	var (
+		dbGroupInfo []sqlc.KeyGroupInfoView
+		groups      []*asset.AssetGroup
+		err         error
+	)
+
+	dbGroupInfo, err = q.AllGroups(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("unable to fetch asset groups: %w", err)
+	}
+
+	// Look up matching genesis info for each group. Each group
+	// should have genesis info populated.
+	for _, g := range dbGroupInfo {
+		groupGenesis, err := fetchGenesis(
+			ctx, q, g.GenAssetID,
+		)
+		if err != nil {
+			return nil, fmt.Errorf("unable to fetch group "+
+				"genesis info: %w", err)
+		}
+
+		groupKey, err := parseGroupInfo(g)
+		if err != nil {
+			return nil, err
+		}
+
+		groups = append(groups, &asset.AssetGroup{
+			Genesis:  &groupGenesis,
+			GroupKey: groupKey,
+		})
+	}
+
+	return groups, nil
+}
+
+// parseGroupInfo maps information on a key group into a GroupKey.
+func parseGroupInfo(group sqlc.KeyGroupInfoView) (*asset.GroupKey, error) {
+	genesisSig, err := schnorr.ParseSignature(group.GenesisSig)
+	if err != nil {
+		return nil, err
+	}
+
+	tweakedGroupKey, err := btcec.ParsePubKey(group.TweakedGroupKey)
+	if err != nil {
+		return nil, err
+	}
+
+	untweakedKey, err := btcec.ParsePubKey(group.RawKey)
```

Comment:
Ah, so this is never nil even for non-local keys, it's just the same value as the tweaked key?

Reply:
Actually we could make it nil-able, I think to do so we'd need to:

Change `asset_groups.internal_key_id` and remove the `NOT NULL` constraint
https://github.com/lightninglabs/taro/blob/ddb0e0d0c9f73800b94b710d4b279e2f0fa5dc72/tarodb/sqlc/migrations/000002_assets.up.sql#L83

(I think) use a `LEFT JOIN` in `key_group_info_view` so that groups without an internal key are not filtered out
https://github.com/lightninglabs/taro/blob/ddb0e0d0c9f73800b94b710d4b279e2f0fa5dc72/tarodb/sqlc/migrations/000002_assets.up.sql#L285

And likely update the CTEs that recreate that view

@Roasbeef 

My understanding of #91 is that anyone receiving would then import the true raw key and this would indeed break local vs. foreign classification here.

Maybe we should use a different table for imported keys altogether? Or have nil-able key_family and key_index.

Reply:
Another option here is to use the `internal_keys.key_family` value to distinguish between local and foreign keys; a local key MUST have the Taro-specific `key_family` of 219, while a foreign key would be imported without any family or index info, so should have a `key_family` of 0.

Reply:
> Another option here is to use the internal_keys.key_family value to distinguish between local and foreign keys

Yeah this can work, or even just make then NULLABLE. So then we'd be able to decouple a key from its locator either via that, or explicitly.